File: Array Diff/Array Diff.java


public class ArrayDiff {
    // Time Complexity: O(n + m), where n = arr1.length, m = arr2.length
    public static int[] arrayDiff(int[] arr1, int[] arr2) {
        java.util.Set<Integer> set1 = new java.util.HashSet<>();
        java.util.Set<Integer> set2 = new java.util.HashSet<>();
        for (int i : arr1) set1.add(i);
        for (int i : arr2) set2.add(i);
        java.util.Set<Integer> diff = new java.util.HashSet<>(set1);
        diff.removeAll(set2);
        for (int i : set2) if (!set1.contains(i)) diff.add(i);
        return diff.stream().sorted().mapToInt(Integer::intValue).toArray();
    }
    public static void main(String[] args) {
        int[] arr1 = {1, 2, 3};
        int[] arr2 = {3, 4, 5};
        int[] result = arrayDiff(arr1, arr2);
        System.out.print("[");
        for (int i = 0; i < result.length; i++) {
            System.out.print(result[i] + (i < result.length - 1 ? ", " : ""));
        }
        System.out.println("]");
    }
}

File: Array Diff/Array Diff.c


#include <stdio.h>
#include <stdlib.h>

// Time Complexity: O(n1 + n2)
void array_diff(int *arr1, int n1, int *arr2, int n2) {
    int *result = malloc((n1 + n2) * sizeof(int));
    int k = 0;
    for (int i = 0; i < n1; i++) {
        int found = 0;
        for (int j = 0; j < n2; j++) {
            if (arr1[i] == arr2[j]) {
                found = 1;
                break;
            }
        }
        if (!found) result[k++] = arr1[i];
    }
    for (int i = 0; i < n2; i++) {
        int found = 0;
        for (int j = 0; j < n1; j++) {
            if (arr2[i] == arr1[j]) {
                found = 1;
                break;
            }
        }
        if (!found) result[k++] = arr2[i];
    }
    // Sort result
    for (int i = 0; i < k - 1; i++) {
        for (int j = i + 1; j < k; j++) {
            if (result[i] > result[j]) {
                int temp = result[i];
                result[i] = result[j];
                result[j] = temp;
            }
        }
    }
    printf("[");
    for (int i = 0; i < k; i++) {
        printf("%d%s", result[i], i < k - 1 ? ", " : "");
    }
    printf("]\n");
    free(result);
}

int main() {
    int arr1[] = {1, 2, 3};
    int arr2[] = {3, 4, 5};
    array_diff(arr1, 3, arr2, 3);
    return 0;
}
